module/zfs/dmu_tx.c

```diff
@@ -1215,13 +1213,74 @@ dmu_tx_assign(dmu_tx_t *tx, uint64_t flags)
 	if ((flags & DMU_TX_NOTHROTTLE))
 		tx->tx_dirty_delayed = B_TRUE;
 
-	while ((err = dmu_tx_try_assign(tx, flags)) != 0) {
+	if (!(flags & DMU_TX_SUSPEND))
+		tx->tx_break_on_suspend = B_TRUE;
+
+	while ((err = dmu_tx_try_assign(tx)) != 0) {
 		dmu_tx_unassign(tx);
 
+		boolean_t suspended = (err == ESHUTDOWN);
+		if (suspended) {
+			/*
+			 * Pool suspended. We need to decide whether to block
+			 * and retry, or return error, depending on the
+			 * caller's flags and the pool config.
+			 */
+			if (flags & DMU_TX_SUSPEND)
+				/*
+				 * The caller expressly does not care about
+				 * suspend, so treat it as a normal retry.
+				 */
+				err = SET_ERROR(ERESTART);
+			else if ((flags & DMU_TX_WAIT) &&
+			    spa_get_failmode(tx->tx_pool->dp_spa) ==
+			    ZIO_FAILURE_MODE_CONTINUE)
+				/*
+				 * Caller wants to wait, but pool config is
+				 * overriding that, so return EIO to be
+				 * propagated back to userspace.
+				 */
+				err = SET_ERROR(EIO);
+			else
+				/* Anything else, we should just block. */
+				err = SET_ERROR(ERESTART);
```

Comment:
```suggestion
				return (SET_ERROR(EIO));
			else
				return (SET_ERROR(ERESTART));
```
Otherwise I am not sure how do we get out of the tight loop.  But I also wonder what reaction do we expect on ERESTART from VFS or user-space?  If they actually restart, we'll again get a tight loop.

Reply:
We don't ever want to return `ERESTART` for `DMU_TX_WAIT`. The existing return check takes care of that. It also handles the `EIO` for `failmode=continue`.

```c
		if (err != ERESTART || !(flags & DMU_TX_WAIT))
			return (err);
```

So far, this isn't really different to the original code, except that I've moved the failmode check down from `dmu_tx_try_assign()`.

So, if the pool suspended and we're not interested (`DMU_TX_SUSPEND` or `failmode=wait`), `err == ERESTART` and we keep moving. `suspended` is true, so we indicate that we want to block in `dmu_tx_wait()` if the pool is suspended, and call it.

```c
		if (suspended)
			tx->tx_break_on_suspend = B_FALSE;

		dmu_tx_wait(tx);
```

In `dmu_tx_wait`:

```c
	txg_wait_flag_t flags =
	    (tx->tx_break_on_suspend ? TXG_WAIT_SUSPEND : TXG_WAIT_NONE);
```

(remember, `TXG_WAIT_SUSPEND` has the inverse meaning to `DMU_TX_SUSPEND`: it will block if the pool is suspended)

So we block until the pool unsuspends, then return to `dmu_tx_assign()`. `tx_break_on_suspend` is reset for `TXG_WAIT_SUSPEND`, and we loop around to `dmu_tx_try_assign()` again.

So I don't think there's any difference in behaviour, because it will either return `EIO` on suspend, or it will block in `txg_wait_synced_flags()`, just as it did before.